pkg/apiserver/rest/dualwriter_mode3.go

```diff
@@ -201,87 +198,83 @@ func (d *DualWriterMode3) Delete(ctx context.Context, name string, deleteValidat
+			d.recordStorageDuration(true, mode3Str, d.resource, method, startStorage)
+		}
+		return objFromStorage, asyncStorage, err
 	}
+	d.recordStorageDuration(false, mode3Str, d.resource, method, startStorage)
+
 	areEqual := Compare(objFromStorage, objFromLegacy)
 	d.recordOutcome(mode3Str, name, areEqual, method)
 	if !areEqual {
-		log.Info("object from legacy and storage are not equal")
+		log.WithValues("name", name).Info("object from legacy and storage are not equal")
 	}
 
-	return err
+	return objFromStorage, asyncStorage, err
 }
 
 // Update overrides the behavior of the generic DualWriter and writes first to Storage and then to LegacyStorage.
 func (d *DualWriterMode3) Update(ctx context.Context, name string, objInfo rest.UpdatedObjectInfo, createValidation rest.ValidateObjectFunc, updateValidation rest.ValidateObjectUpdateFunc, forceAllowCreate bool, options *metav1.UpdateOptions) (runtime.Object, bool, error) {
 	var method = "update"
 	log := d.Log.WithValues("name", name, "method", method)
 	ctx = klog.NewContext(ctx, log)
-
-	startStorage := time.Now()
-	objFromStorage, async, err := d.Storage.Update(ctx, name, objInfo, createValidation, updateValidation, forceAllowCreate, options)
-	d.recordStorageDuration(err != nil, mode3Str, d.resource, method, startStorage)
-	if err != nil {
-		log.Error(err, "unable to update in storage")
-		return objFromStorage, async, err
-	}
-
-	//nolint:errcheck
-	go d.updateOnLegacyStorageMode3(ctx, objFromStorage, name, objInfo, createValidation, updateValidation, forceAllowCreate, options)
-
-	return objFromStorage, async, err
-}
-
-func (d *DualWriterMode3) updateOnLegacyStorageMode3(ctx context.Context, storageObj runtime.Object, name string, objInfo rest.UpdatedObjectInfo, createValidation rest.ValidateObjectFunc, updateValidation rest.ValidateObjectUpdateFunc, forceAllowCreate bool, options *metav1.UpdateOptions) error {
-	// The incoming RV is from unified storage, so legacy can ignore it
+	// The incoming RV is not stable -- it may be from legacy or storage!
+	// This sets a flag in the context and our apistore is more lenient when it exists
 	ctx = context.WithValue(ctx, dualWriteContextKey{}, true)
 
-	var method = "update"
-	log := d.Log.WithValues("name", name, "method", method, "name", name)
+	// update in legacy first, and then unistore. Will return a failure if either fails.
+	//
+	// we want to update in legacy first, otherwise if the update from unistore was successful,
+	// but legacy failed, the user would get a failure, but see the update did apply to the source
+	// of truth, and be less likely to retry to save (and get the stores in sync again)
 
-	ctx, cancel := context.WithTimeoutCause(context.WithoutCancel(ctx), time.Second*10, errors.New("legacy update timeout"))
 	startLegacy := time.Now()
-	defer cancel()
+	objFromLegacy, createdLegacy, err := d.Legacy.Update(ctx, name, objInfo, createValidation, updateValidation, forceAllowCreate, options)
+	if err != nil {
```

Comment:
@leonorfmartins I am trying to remember if some of the metatada (eg uuids) will become inconstistent if we flip the order here. Do you remember ?

Reply:
🤔 the issue might be on create, actually. 
I think it's better to test it, but I remember something about unistore having a rule on the storage backend side of generating a name if one wasn't present: but I dont think we have that on legacy side, meaning, legacy will always generate an UID. Which means, if we generate a name on unistore first, that one will be ignored when we create the same object in legacy.
We can change the logic on legacy, though

Reply:
We should be all set:
https://github.com/user-attachments/assets/3d1164b0-b4d6-4171-a808-14dbd1110bda